File: CBHTTP/Connectivity/ConnectionStatus.swift
// Copyright (c) 2017-2018 Coinbase Inc. See LICENSE

import SystemConfiguration

/// Represents network status
public enum ConnectionStatus: Equatable {
    /// Network is online using the given network kind i.e. wifi, wwan, etc
    case connected(ConnectionKind)

    /// Network is offline
    case offline

    /// Network status is unknown
    case unknown

    /// Network is online
    public var isOnline: Bool {
        guard case .connected = self else { return false }
        return true
    }

    init(flags: SCNetworkReachabilityFlags) {
        if flags.isNetworkReachable {
            // Currently online
            self = flags.contains(.isWWAN) ? .connected(.wwan) : .connected(.wifi)
        } else {
            // Currently offline
            self = .offline
        }
    }
}

File: android/http/src/main/java/com/coinbase/wallet/http/connectivity/ConnectionStatus.kt
package com.coinbase.wallet.http.connectivity

/**
 * Represents data type received via web live connection
 */
sealed class ConnectionStatus {
    /**
     * Network is online using the given network kind i.e. wifi, wwan, etc
     */
    data class Connected(val kind: ConnectionKind) : ConnectionStatus()

    /**
     * Network is offline
     */
    object Offline : ConnectionStatus()

    /**
     * Network status is unknown
     */
    object Unknown : ConnectionStatus()

    /**
     * Network is online
     */
    val isOnline: Boolean get() = this is Connected
}
